core/trino-main/src/main/java/io/trino/sql/planner/PartitioningProviderManager.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner;
+
+import io.trino.Session;
+import io.trino.connector.CatalogName;
+import io.trino.spi.connector.BucketFunction;
+import io.trino.spi.connector.ConnectorBucketNodeMap;
+import io.trino.spi.connector.ConnectorNodePartitioningProvider;
+import io.trino.spi.type.Type;
+
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class PartitioningProviderManager
+{
+    private final ConcurrentMap<CatalogName, ConnectorNodePartitioningProvider> partitioningProviders = new ConcurrentHashMap<>();
+
+    public ConnectorBucketNodeMap getConnectorBucketNodeMap(Session session, PartitioningHandle partitioning)
+    {
+        ConnectorNodePartitioningProvider partitioningProvider = getPartitioningProvider(partitioning.getConnectorId().get());
+        ConnectorBucketNodeMap connectorBucketNodeMap = partitioningProvider.getBucketNodeMap(
+                partitioning.getTransactionHandle().orElse(null),
+                session.toConnectorSession(partitioning.getConnectorId().get()),
+                partitioning.getConnectorHandle());
+        checkArgument(connectorBucketNodeMap != null, "No partition map %s", partitioning);
+        return connectorBucketNodeMap;
+    }
+
+    public BucketFunction getBucketFunction(Session session, PartitioningHandle partitioning, List<Type> partitionChannelTypes, int bucketCount)
+    {
+        ConnectorNodePartitioningProvider partitioningProvider = getPartitioningProvider(partitioning.getConnectorId().get());
+        BucketFunction bucketFunction = partitioningProvider.getBucketFunction(
+                partitioning.getTransactionHandle().orElse(null),
+                session.toConnectorSession(),
+                partitioning.getConnectorHandle(),
+                partitionChannelTypes,
+                bucketCount);
+        checkArgument(bucketFunction != null, "No bucket function for partitioning: %s", partitioning);
+        return bucketFunction;
+    }
+
+    public ConnectorNodePartitioningProvider getPartitioningProvider(CatalogName catalogName)
```

Comment:
Please wrap other methods from `ConnectorNodePartitioningProvider` and drop `getPartitioningProvider`.

Reply:
I actually got rid of entire `PartitioningProviderManager` as split between `PartitioningProviderManager` and `NodePartitioningManager` was non-obvious.